ledger/accountdb.go

```diff
@@ -215,9 +279,214 @@ func prepareNormalizedBalances(bals []encodedBalanceRecord, proto config.Consens
 	return
 }
 
+// makeCompactResourceDeltas takes an array of AccountDeltas ( one array entry per round ), and compacts the resource portions of the arrays into a single
+// data structure that contains all the resources deltas changes. While doing that, the function eliminate any intermediate resources changes.
+// It counts the number of changes each account get modified across the round range by specifying it in the nAcctDeltas field of the accountDeltaCount/modifiedCreatable.
+func makeCompactResourceDeltas(accountDeltas []ledgercore.NewAccountDeltas, baseResources lruResources) (outResourcesDeltas compactResourcesDeltas) {
+	if len(accountDeltas) == 0 {
+		return
+	}
+
+	// the sizes of the maps here aren't super accurate, but would hopefully be a rough estimate for a reasonable starting point.
+	size := accountDeltas[0].Len()*len(accountDeltas) + 1
+	outResourcesDeltas.cache = make(map[accountCreatable]int, size)
+	outResourcesDeltas.deltas = make([]resourcesDeltas, 0, size)
+	outResourcesDeltas.misses = make([]int, 0, size)
+
+	for _, roundDelta := range accountDeltas {
+		// assets
+		for _, assetHold := range roundDelta.GetAllAssetsHoldings() {
+			if prev, idx := outResourcesDeltas.get(assetHold.Addr, basics.CreatableIndex(assetHold.Aidx)); idx != -1 {
+				// update existing entry with new data.
+				updEntry := resourcesDeltas{
+					oldResource: prev.oldResource,
+					nAcctDeltas: prev.nAcctDeltas + 1,
+				}
+				if assetHold.Holding != nil {
+					updEntry.newResource.SetAssetHolding(*assetHold.Holding)
+				} else {
+					updEntry.newResource.ClearAssetHolding()
+				}
+				outResourcesDeltas.update(idx, updEntry)
+			} else {
+				// it's a new entry.
+				newEntry := resourcesDeltas{
+					nAcctDeltas: 1,
+				}
+				if assetHold.Holding != nil {
+					newEntry.newResource.SetAssetHolding(*assetHold.Holding)
+				} else {
+					newEntry.newResource.ClearAssetHolding()
+				}
+				if baseResourceData, has := baseResources.read(assetHold.Addr, basics.CreatableIndex(assetHold.Aidx)); has {
+					newEntry.oldResource = baseResourceData
+					outResourcesDeltas.insert(assetHold.Addr, basics.CreatableIndex(assetHold.Aidx), newEntry) // insert instead of upsert economizes one map lookup
+				} else {
+					newEntry.oldResource.data.ClearAssetHolding()
```

Comment:
this is a new entry, no clearing needed

Reply:
yes, I was afraid that you'll say that. I should probably document this better.
The reason I added `ClearAssetHolding()` is so that it would set the `resourceFlagsNotHolding` flag.

Reply:
we discussed this offline.